File: web/src/lib/api/currency.ts
import { request } from './client'
import type { CurrencyRatesResponse } from './generated'

export type { CurrencyRatesResponse } from './generated'

/** Latest cached reference rates, expressed as units of each currency per USD. */
export function getCurrencyRates(): Promise<CurrencyRatesResponse> {
  return request<CurrencyRatesResponse>('/api/currencies')
}

File: api/src/handlers/currency.rs
use axum::{Json, extract::State};

use crate::{currency::CurrencyRatesResponse, error::AppError, state::AppState};

/// `GET /api/currencies` -> the cached latest USD reference rates used by the SPA for
/// display-only conversion. Catalog values remain USD on the wire and in storage.
pub async fn currency_rates(
    State(state): State<AppState>,
) -> Result<Json<CurrencyRatesResponse>, AppError> {
    Ok(Json(state.currency_rates.latest(&state.http).await?))
}
